File: September2020/MaximumProductSubArray.java
class MaximumProductSubArray {
    // Space => O(1), Time Complexity => O(n)
    public int maxProduct(int[] nums) {
        int maxSoFar = nums[0];
        int minSoFar = nums[0];
        int res = nums[0];
        for(int i=1;i<nums.length;i++){
            int copyMaxSoFar = maxSoFar;
            int el = nums[i];
            maxSoFar = Math.max(el, Math.max(el*maxSoFar, el*minSoFar));

            minSoFar = Math.min(el, Math.min(el*copyMaxSoFar, el*minSoFar));
            res = Math.max(res, maxSoFar);
        }
        return res;
    }
}

File: December2021/C++/Maximum_Product_Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int max_so_far=nums[0],curr_max=nums[0],curr_min=nums[0];
        for(int i=1;i<nums.size();i++)
        {
            int temp=curr_max;
            curr_max=max({nums[i],temp*nums[i],curr_min*nums[i]});
            curr_min=min({nums[i],temp*nums[i],curr_min*nums[i]});
            max_so_far=max({max_so_far,curr_max,curr_min});
        }
        return max_so_far;
    }
};